packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -11,10 +11,707 @@ import 'picker.dart';
+      _getPickerWidth('h'),
+      _getPickerWidth('m'),
+    ];
+    final List<Function> pickerBuilder = <Function>[
+      _buildHourPicker,
+      _buildMinutePicker,
+    ];
+
+    // Adds am/pm column if the picker is not using 24h format.
+    if (!widget.use24hFormat) {
+      if (localizations.datePickerDateTimeOrder == DatePickerDateTimeOrder.date_time_dayPeriod
+        || localizations.datePickerDateTimeOrder == DatePickerDateTimeOrder.time_dayPeriod_date) {
+        pickerBuilder.add(_buildAmPmPicker);
+        columnWidth.add(_getPickerWidth('a'));
+      }
+      else {
+        pickerBuilder.insert(0, _buildAmPmPicker);
+        columnWidth.insert(0, _getPickerWidth('a'));
+      }
+    }
+
+    // Adds medium date column if the picker's mode is date and time.
+    if (widget.mode == CupertinoDatePickerMode.dateAndTime) {
+      if (localizations.datePickerDateTimeOrder == DatePickerDateTimeOrder.time_dayPeriod_date
+          || localizations.datePickerDateTimeOrder == DatePickerDateTimeOrder.dayPeriod_time_date) {
+        pickerBuilder.add(_buildMediumDatePicker);
+        columnWidth.add(_getPickerWidth('d'));
+      }
+      else {
+        pickerBuilder.insert(0, _buildMediumDatePicker);
+        columnWidth.insert(0, _getPickerWidth('d'));
+      }
+    }
+
+    final List<Widget> children = <Widget>[];
+
+    for (int i = 0; i < columnWidth.length; i++) {
+      double offAxisFraction = 0.0;
+      if (i == 0)
+        offAxisFraction = -0.5 * textDirectionFactor;
+      else if (i >= 2 || columnWidth.length == 2)
+        offAxisFraction = 0.5 * textDirectionFactor;
+
+      EdgeInsets padding = const EdgeInsets.only(right: _kDatePickerPadSize);
+      if (i == columnWidth.length - 1)
+        padding = padding.flipped;
+      if (textDirectionFactor == -1)
+        padding = padding.flipped;
+
+      children.add(LayoutId(
+        id: i,
+        child: pickerBuilder[i](
+          offAxisFraction,
+          (Widget child) {
+            return new Container(
+              alignment: i == columnWidth.length - 1
+                ? alignCenterLeft
+                : alignCenterRight,
+              padding: padding,
+              child: new Container(
+                alignment: i == columnWidth.length - 1 ? alignCenterLeft : alignCenterRight,
+                width: i == 0 || i == columnWidth.length - 1
+                  ? null
+                  : columnWidth[i] + _kDatePickerPadSize,
+                child: child,
+              ),
+            );
+          },
+        ),
+      ));
+    }
+
+    return new MediaQuery(
+      data: const MediaQueryData(textScaleFactor: 1.0),
+      child: new CustomMultiChildLayout(
+        delegate: new _DatePickerLayout(
+          columnWidth: columnWidth,
+          textDirectionFactor: textDirectionFactor,
+        ),
+        children: children,
+      ),
+    );
+  }
+}
+
+class _CupertinoDatePickerDateState extends State<CupertinoDatePicker> {
```

Comment:
I feel like a number of things can be shared. Can't they be extracted into a mixin or a superclass?

Reply:
From the users' view, I want them to see this as just one class (same as how iOS does), and mostly what users actually need to do is specifying the mode.
I've thought about writing superclass and inheritance, but that will also goes to user's level, and then calling a picker will goes to new class names like CupertinoDatePicker, CupertinoDateTimePicker, CupertinoTimePicker, etc.

Reply:
Sorry, I meant extracting common logic between the private _CupertinoDatePickerDateTimeState and _CupertinoDatePickerDateState. They could be private static functions in this file too.

Reply:
The one function that is I think can be shared is `_getColumnWidth`, however that one is reading information from the current BuildContext, so I can't extract that out. Any suggestion?

Reply:
You can pass this.context as a parameter to static functions.